File: Proyecto_SOA-master/backend-products/app/controllers/sales_controller.py
# app/controllers/sales_controller.py
from flask import request, jsonify
from decimal import Decimal
from app.config.database import db
from mysql.connector import Error

def create_sale():
    try:
        data = request.get_json()
        items = data.get('items')

        if not items or len(items) == 0:
            return jsonify({'success': False, 'message': 'Debe incluir productos en la venta'}), 400

        seller_name = data.get('seller_name')
        seller_identifier = data.get('seller_identifier')
        customer_name = data.get('customer_name')
        customer_email = data.get('customer_email')
        payment_method = data.get('payment_method', 'efectivo')

        connection = db.get_connection()
        cursor = connection.cursor(dictionary=True)

        total_amount = Decimal('0.00')
        products_data = {}

        for item in items:
            product_id = item.get('product_id')
            quantity = item.get('quantity')

            cursor.execute("SELECT id, name, price, stock FROM products WHERE id = %s", (product_id,))
            product = cursor.fetchone()

            if not product:
                return jsonify({'success': False, 'message': f'Producto {product_id} no encontrado'}), 404

            if product['stock'] < quantity:
                return jsonify({
                    'success': False,
                    'message': f"Stock insuficiente para {product['name']}. Disponible: {product['stock']}, solicitado: {quantity}"
                }), 400

            unit_price = Decimal(str(product['price']))
            subtotal = unit_price * quantity
            total_amount += subtotal

            products_data[product_id] = {
                "name": product["name"],
                "unit_price": unit_price,
                "quantity": quantity,
                "subtotal": subtotal
            }

        cursor.execute("""
            INSERT INTO sales 
            (seller_name, seller_identifier, customer_name, customer_email, total_amount, payment_method)
            VALUES (%s, %s, %s, %s, %s, %s)
        """, (seller_name, seller_identifier, customer_name, customer_email, float(total_amount), payment_method))

        sale_id = cursor.lastrowid

        for pid, info in products_data.items():
            cursor.execute("""
                INSERT INTO sale_items 
                (sale_id, product_id, product_name, unit_price, quantity, subtotal)
                VALUES (%s, %s, %s, %s, %s, %s)
            """, (
                sale_id, pid, info["name"], float(info["unit_price"]), info["quantity"], float(info["subtotal"])
            ))

            cursor.execute("""
                UPDATE products SET stock = stock - %s WHERE id = %s
            """, (info["quantity"], pid))

        connection.commit()
        cursor.close()

        return jsonify({
            "success": True,
            "message": "Venta registrada con éxito",
            "sale_id": sale_id,
            "total": float(total_amount)
        }), 201

    except Exception as e:
        return jsonify({'success': False, 'message': f'Error: {str(e)}'}), 500
def get_sales_history():
    seller_identifier = request.args.get("seller_identifier")  # filtrado por vendedor opcional

    connection = db.get_connection()
    cursor = connection.cursor(dictionary=True)

    query = """
        SELECT id, seller_name, seller_identifier, customer_name, total_amount, payment_method, created_at
        FROM sales
        WHERE 1 = 1
    """
    params = []

    # FILTROS OPCIONALES 🔥
    if seller_identifier:
        query += " AND seller_identifier = %s"
        params.append(seller_identifier)

    start = request.args.get("start_date")
    end   = request.args.get("end_date")
    client = request.args.get("customer_name")

    if start:
        query += " AND DATE(created_at) >= %s"
        params.append(start)

    if end:
        query += " AND DATE(created_at) <= %s"
        params.append(end)

    if client:
        query += " AND customer_name LIKE %s"
        params.append(f"%{client}%")

    query += " ORDER BY created_at DESC"   # ⬅ ordenado del más reciente al más viejo

    cursor.execute(query, params)
    sales = cursor.fetchall()
    cursor.close()

    return jsonify({"success": True, "sales": sales}), 200


# ────────────────────────────────────────────────
#  🔥 DETALLE DE UNA VENTA
# ────────────────────────────────────────────────

def get_sale_details(sale_id):
    connection = db.get_connection()
    cursor = connection.cursor(dictionary=True)

    cursor.execute("""
        SELECT id, seller_name, seller_identifier, customer_name, customer_email,
               total_amount, payment_method, created_at
        FROM sales
        WHERE id = %s
    """, (sale_id,))
    sale = cursor.fetchone()

    if not sale:
        return jsonify({"success": False, "message": "Venta no encontrada"}), 404

    cursor.execute("""
        SELECT product_id, product_name, unit_price, quantity, subtotal
        FROM sale_items
        WHERE sale_id = %s
    """, (sale_id,))
    sale["items"] = cursor.fetchall()
    cursor.close()

    return jsonify({"success": True, "sale": sale}), 200

File: Proyecto_SOA-master/backend-users/src/controllers/salesController.js
// src/controllers/salesController.js
const Sale = require('../models/Sale');
const axios = require('axios');

// URL del servicio de productos (Flask/Python)
const PRODUCTS_SERVICE_URL = process.env.PRODUCTS_SERVICE_URL || 'http://localhost:5001';

// @desc    Registrar nueva venta
// @route   POST /api/sales
// @access  Private/Vendedor
exports.createSale = async (req, res) => {
  try {
    const { customerName, customerEmail, customerPhone, items, paymentMethod } = req.body;

    // Validar que vengan items
    if (!items || items.length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Debe agregar al menos un producto a la venta'
      });
    }

    // Validar cliente
    if (!customerName) {
      return res.status(400).json({
        success: false,
        message: 'El nombre del cliente es requerido'
      });
    }

    // 1. Verificar stock disponible para cada producto
    for (const item of items) {
      try {
        const productResponse = await axios.get(
          `${PRODUCTS_SERVICE_URL}/api/products/${item.productId}`
        );

        const product = productResponse.data;

        if (!product || !product.id) {
          return res.status(404).json({
            success: false,
            message: `Producto con ID ${item.productId} no encontrado`
          });
        }

        if (product.stock < item.quantity) {
          return res.status(400).json({
            success: false,
            message: `Stock insuficiente para ${product.name}. Disponible: ${product.stock}, Solicitado: ${item.quantity}`
          });
        }

        // Agregar datos del producto al item
        item.productName = product.name;
        item.unitPrice = product.price;
        item.subtotal = product.price * item.quantity;

      } catch (error) {
        console.error('Error al verificar producto:', error.message);
        return res.status(500).json({
          success: false,
          message: `Error al verificar disponibilidad del producto ${item.productId}`
        });
      }
    }

    // 2. Calcular total
    const totalAmount = items.reduce((sum, item) => sum + item.subtotal, 0);

    // 3. Crear la venta en MongoDB
    const sale = await Sale.create({
      sellerId: req.user._id,
      customerName,
      customerEmail,
      customerPhone,
      items,
      totalAmount,
      paymentMethod: paymentMethod || 'efectivo',
      status: 'completada',
      saleDate: new Date()
    });

    // 4. Actualizar stock de cada producto
    for (const item of items) {
      try {
        await axios.put(
          `${PRODUCTS_SERVICE_URL}/api/products/${item.productId}/stock`,
          { quantity: item.quantity }
        );
      } catch (error) {
        console.error(`Error al actualizar stock del producto ${item.productId}:`, error.message);
        
        // IMPORTANTE: Si falla la actualización de stock, cancelar la venta
        await Sale.findByIdAndUpdate(sale._id, { status: 'cancelada' });
        
        return res.status(500).json({
          success: false,
          message: 'Error al actualizar inventario. Venta cancelada.'
        });
      }
    }

    console.log(`✅ Venta registrada: ${sale._id} - Total: $${totalAmount}`);

    res.status(201).json({
      success: true,
      message: 'Venta registrada exitosamente',
      data: sale
    });

  } catch (error) {
    console.error('Error al registrar venta:', error);
    
    if (error.name === 'ValidationError') {
      const messages = Object.values(error.errors).map(err => err.message);
      return res.status(400).json({
        success: false,
        message: messages.join(', ')
      });
    }

    res.status(500).json({
      success: false,
      message: 'Error al registrar la venta'
    });
  }
};

// @desc    Obtener historial de ventas del vendedor
// @route   GET /api/sales/my-sales
// @access  Private/Vendedor
exports.getMySales = async (req, res) => {
  try {
    const { startDate, endDate, status } = req.query;

    // Construir filtro
    let filter = { sellerId: req.user._id };

    if (status) {
      filter.status = status;
    }

    if (startDate || endDate) {
      filter.saleDate = {};
      if (startDate) filter.saleDate.$gte = new Date(startDate);
      if (endDate) filter.saleDate.$lte = new Date(endDate);
    }

    const sales = await Sale.find(filter)
      .sort({ saleDate: -1 })
      .limit(100);

    const totalAmount = await Sale.aggregate([
      { $match: filter },
      { $group: { _id: null, total: { $sum: '$totalAmount' } } }
    ]);

    res.status(200).json({
      success: true,
      count: sales.length,
      totalAmount: totalAmount[0]?.total || 0,
      data: sales
    });

  } catch (error) {
    console.error('Error al obtener ventas:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener historial de ventas'
    });
  }
};

// @desc    Obtener todas las ventas (Admin/Consultor)
// @route   GET /api/sales
// @access  Private/Admin/Consultor
exports.getAllSales = async (req, res) => {
  try {
    const { sellerId, startDate, endDate, status } = req.query;

    let filter = {};

    if (sellerId) filter.sellerId = sellerId;
    if (status) filter.status = status;

    if (startDate || endDate) {
      filter.saleDate = {};
      if (startDate) filter.saleDate.$gte = new Date(startDate);
      if (endDate) filter.saleDate.$lte = new Date(endDate);
    }

    const sales = await Sale.find(filter)
      .populate('sellerId', 'name email')
      .sort({ saleDate: -1 });

    const totalAmount = await Sale.aggregate([
      { $match: filter },
      { $group: { _id: null, total: { $sum: '$totalAmount' } } }
    ]);

    res.status(200).json({
      success: true,
      count: sales.length,
      totalAmount: totalAmount[0]?.total || 0,
      data: sales
    });

  } catch (error) {
    console.error('Error al obtener ventas:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener ventas'
    });
  }
};

// @desc    Obtener venta por ID
// @route   GET /api/sales/:id
// @access  Private
exports.getSaleById = async (req, res) => {
  try {
    const sale = await Sale.findById(req.params.id)
      .populate('sellerId', 'name email phone');

    if (!sale) {
      return res.status(404).json({
        success: false,
        message: 'Venta no encontrada'
      });
    }

    // Vendedores solo pueden ver sus propias ventas
    if (req.user.role === 'vendedor' && 
        sale.sellerId._id.toString() !== req.user._id.toString()) {
      return res.status(403).json({
        success: false,
        message: 'No tienes permiso para ver esta venta'
      });
    }

    res.status(200).json({
      success: true,
      data: sale
    });

  } catch (error) {
    console.error('Error al obtener venta:', error);
    res.status(500).json({
      success: false,
      message: 'Error al obtener venta'
    });
  }
};